pulp_container/app/tasks/sign.py

```diff
@@ -0,0 +1,122 @@
+        if tagged_manifest.media_type in MANIFEST_MEDIA_TYPES.IMAGE:
+            signature_pk = create_signature(
+                tagged_manifest, ":".join((reference, tag.name)), signing_service
+            )
+            added_signatures.append(signature_pk)
+        elif tagged_manifest.media_type in MANIFEST_MEDIA_TYPES.LIST:
+            # parse ML and sign per-arches by digest
+            for manifest in tagged_manifest.manifests:
+                signature_pk = create_signature(
+                    manifest, ":".join((reference, manifest.digest)), signing_service
+                )
+                added_signatures.append(signature_pk)
+
+    added_signatures_qs = ManifestSignature.objects.filter(pk__in=added_signatures)
+    with repository.new_version() as new_version:
+        new_version.add_content(added_signatures_qs)
+
+
+def create_signature(manifest, reference, signing_service):
+    """
+    Creates manifest signature.
+
+    Created signature is extracted,parsed and a ManifestSignature is saved.
+
+    Args:
+        manifest (models.Manifest): Manifest which is intended to be signed.
+        reference (str): reference that will be used in signature's docker-reference.
+        signing_service (models.ManifestSigingService): signing service that will be used.
+
+    Returns:
+        pk of created ManifestSignature.
+
+    """
+    with tempfile.TemporaryDirectory(".") as working_directory:
+        # download and write file for object storage
+        if settings.DEFAULT_FILE_STORAGE != "pulpcore.app.models.storage.FileSystem":
+            manifest_file = tempfile.NamedTemporaryFile(dir=working_directory, delete=False)
+            manifest_file.write(manifest._artifacts.get().file.read())
+            manifest_file.flush()
+            manifest_path = manifest_file.name
+        else:
+            manifest_path = manifest._artifacts.get().file.path
+        sig_path = os.path.join(working_directory, "signature")
+
+        signed = signing_service.sign(
+            manifest_path, env_vars={"REFERENCE": reference, "SIG_PATH": sig_path}
+        )
+
+        with open(signed["signature"], "rb") as sig_fp:
+            data = sig_fp.read()
+            encoded_sig = base64.b64encode(data)
+            sig_digest = hashlib.sha256(data).hexdigest()
+            sig_json = extract_data_from_signature(data, manifest.digest)
+            manifest_digest = sig_json["critical"]["image"]["docker-manifest-digest"]
+
+            signature = ManifestSignature(
+                name=f"{manifest_digest}@{sig_digest[:32]}",
+                digest=f"sha256:{sig_digest}",
+                type=SIGNATURE_TYPE.ATOMIC_SHORT,
+                key_id=sig_json["signing_key_id"],
+                timestamp=sig_json["signature_timestamp"],
+                creator=sig_json["optional"].get("creator"),
+                data=encoded_sig,
+                signed_manifest=manifest,
+            )
+            signature.save()
```

Comment:
I agree that signature will be different because of changing timestamp.

To be honest I would not go into the business of signature management on sign/re-sign - it user wants to remove some old signatures he should do that explicitly on his own. Even if the signature was produced for same manifest, key id  and docker-reference, the creator can change-  for example on re-sign, signature might have been produced with different version of the signing atomic tool.